homeassistant/components/geo_location/nsw_rural_fire_service_feed.py

```diff
@@ -0,0 +1,244 @@
+            self._update_or_remove_entities([])
+
+    def _update_or_remove_entities(self, feed_entries):
+        """Update existing entries and remove obsolete entities."""
+        _LOGGER.debug("Entries for updating: %s", feed_entries)
+        # Remove obsolete entities for events that have disappeared
+        managed_entities = self._managed_entities.copy()
+        entries_lookup = {entry.external_id: entry for entry in feed_entries}
+        for entity in managed_entities:
+            entry = entries_lookup.pop(entity.external_id, None)
+            if entry is None:
+                # Remove obsolete entity.
+                _LOGGER.debug("Entity not current anymore %s", entity)
+                self._managed_entities.remove(entity)
+                self._hass.add_job(entity.async_remove())
+                continue
+            # Existing entity - update details.
+            _LOGGER.debug("Existing entity found %s", entity)
+            feed_entries.remove(entry)
+            entity.schedule_update_ha_state(True)
+        # Return the remaining entries that new entities must be created for.
+        return feed_entries
+
+    def _generate_new_entities(self, entries):
+        """Generate new entities for events."""
+        new_entities = []
+        for entry in entries:
+            new_entity = NswRuralFireServiceLocationEvent(self, entry)
+            _LOGGER.debug("New entity added %s", new_entity)
+            new_entities.append(new_entity)
+        # Add new entities to HA and keep track of them in this manager.
+        self._add_entities(new_entities, True)
+        self._managed_entities.extend(new_entities)
+
+    def get_feed_entry(self, external_id):
+        """Return a feed entry identified by external id."""
+        return next((entry for entry in self._feed_entries
+                     if entry.external_id == external_id), None)
+
+
+class NswRuralFireServiceLocationEvent(GeoLocationEvent):
+    """This represents an external event with GeoJSON data."""
+
+    def __init__(self, feed_manager, feed_entry):
+        """Initialize entity with data from feed entry."""
+        self._feed_manager = feed_manager
+        self._update_from_feed(feed_entry)
+
+    @property
+    def should_poll(self):
+        """No polling needed for GeoJSON location events."""
+        return False
+
+    async def async_update(self):
+        """Update this entity from the data held in the feed manager."""
+        feed_entry = self._feed_manager.get_feed_entry(self.external_id)
+        if feed_entry:
+            self._update_from_feed(feed_entry)
+
+    def _update_from_feed(self, feed_entry):
+        """Update the internal state from the provided feed entry."""
+        self._name = feed_entry.title
+        self._distance = feed_entry.distance_to_home
+        self._latitude = feed_entry.coordinates[0]
+        self._longitude = feed_entry.coordinates[1]
+        self.external_id = feed_entry.external_id
```

Comment:
We should never change the `external_id` for this entity.

Reply:
Yes - makes much more sense now after refactoring